File: app/src/main/java/com/riders/thelab/ui/contacts/ContactsAdapter.java
package com.riders.thelab.ui.contacts;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.riders.thelab.R;
import com.riders.thelab.data.local.model.Contact;

import java.util.ArrayList;
import java.util.List;

import timber.log.Timber;

public class ContactsAdapter extends RecyclerView.Adapter<ContactsViewHolder>
        implements Filterable {

    private Context context;
    private List<Contact> contactList;
    private List<Contact> contactListFiltered;
    private ContactsClickListener listener;


    public ContactsAdapter(@NonNull Context context,
                           @NonNull ArrayList<Contact> contactList,
                           @NonNull ContactsClickListener listener) {
        this.context = context;
        this.contactList = contactList;
        this.listener = listener;
        this.contactListFiltered = contactList;
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                Timber.d("performFiltering()");
                String charString = charSequence.toString();
                if (charString.isEmpty()) {
                    Timber.d("charString.isEmpty()()");
                    contactListFiltered = contactList;
                } else {
                    List<Contact> filteredList = new ArrayList<>();
                    for (Contact row : contactList) {
                        Timber.d("row : %s", row.toString());

                        // name match condition. this might differ depending on your requirement
                        // here we are looking for name or phone number match
                        if (row.getName().toLowerCase().contains(charString.toLowerCase()) || row.getEmail().contains(charSequence)) {
                            filteredList.add(row);
                        }
                    }

                    contactListFiltered = filteredList;
                    Timber.d("contactListFiltered size : %s", contactListFiltered.size());
                }

                FilterResults filterResults = new FilterResults();
                filterResults.values = contactListFiltered;
                return filterResults;
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                Timber.d("publishResults()");
                contactListFiltered = (ArrayList<Contact>) filterResults.values;

                // refresh the list with filtered data
                notifyDataSetChanged();
            }
        };
    }


    @Override
    public int getItemCount() {
        if (contactListFiltered != null) {
            return contactListFiltered.size();
        }
        return 0;
    }


    @NonNull
    @Override
    public ContactsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ContactsViewHolder(
                context,
                LayoutInflater.from(parent.getContext())
                        .inflate(R.layout.row_contact_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ContactsViewHolder holder, int position) {
        final Contact item = contactListFiltered.get(position);

        holder.bindData(item);
        holder.cardView.setOnClickListener(
                view -> listener.onContactItemCLickListener(item, holder.getAdapterPosition()));
    }


    public void removeItem(int position) {
        contactListFiltered.remove(position);
        // notify the item removed by position
        // to perform recycler view delete animations
        // NOTE: don't call notifyDataSetChanged()
        notifyItemRemoved(position);
    }

    public void restoreItem(Contact item, int position) {
        contactListFiltered.add(position, item);
        // notify item added by position
        notifyItemInserted(position);
    }
}

File: app/src/main/java/com/riders/thelab/ui/contacts/ContactsAdapter.kt
package com.riders.thelab.ui.contacts

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.recyclerview.widget.RecyclerView
import com.riders.thelab.R
import com.riders.thelab.data.local.model.Contact
import timber.log.Timber
import java.util.*

class ContactsAdapter(
        val context: Context,
        val contactList: ArrayList<Contact>,
        val listener: ContactsClickListener
) : RecyclerView.Adapter<ContactsViewHolder>(), Filterable {

    private var contactListFiltered: MutableList<Contact>? = contactList

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContactsViewHolder {
        return ContactsViewHolder(
                context,
                LayoutInflater.from(parent.context)
                        .inflate(R.layout.row_contact_item, parent, false))
    }

    override fun onBindViewHolder(holder: ContactsViewHolder, position: Int) {
        val item = contactListFiltered!![position]

        holder.bindData(item)
        holder.viewBinding.cvContactItem.setOnClickListener { view: View? -> listener.onContactItemCLickListener(item, holder.adapterPosition) }
    }

    override fun getItemCount(): Int {
        return if (contactListFiltered != null) {
            contactListFiltered!!.size
        } else 0
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(charSequence: CharSequence): FilterResults {
                Timber.d("performFiltering()")
                val charString = charSequence.toString()
                if (charString.isEmpty()) {
                    Timber.d("charString.isEmpty()()")
                    contactListFiltered = contactList
                } else {
                    val filteredList: MutableList<Contact> = ArrayList()
                    for (row in contactList) {
                        Timber.d("row : %s", row.toString())

                        // name match condition. this might differ depending on your requirement
                        // here we are looking for name or phone number match
                        if (row.name.toLowerCase(Locale.ROOT).contains(charString.toLowerCase(Locale.ROOT))
                                || row.email.contains(charSequence)) {
                            filteredList.add(row)
                        }
                    }
                    contactListFiltered = filteredList
                    Timber.d("contactListFiltered size : %s", contactListFiltered!!.size)
                }
                val filterResults = FilterResults()
                filterResults.values = contactListFiltered
                return filterResults
            }

            override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
                Timber.d("publishResults()")
                contactListFiltered = filterResults.values as MutableList<Contact>?

                // refresh the list with filtered data
                notifyDataSetChanged()
            }
        }
    }

    fun removeItem(position: Int) {
        contactListFiltered!!.removeAt(position)
        // notify the item removed by position
        // to perform recycler view delete animations
        // NOTE: don't call notifyDataSetChanged()
        notifyItemRemoved(position)
    }

    fun restoreItem(item: Contact, position: Int) {
        contactListFiltered!!.add(position, item)
        // notify item added by position
        notifyItemInserted(position)
    }

}